pkg/runtimeproxy/server/docker/handler.go

```diff
@@ -18,278 +18,46 @@ package docker
+	klog.Errorf("%v", err)
+	http.Error(wr, err.Error(), http.StatusInternalServerError)
 }
 
-func (d *RuntimeManagerDockerServer) HandleStopContainer(ctx context.Context, wr http.ResponseWriter, req *http.Request) {
-	containerID, err := getContainerID(req.URL.Path)
-	if err != nil {
-		klog.Errorf("Failed to get container id, err: %v", err)
-		http.Error(wr, err.Error(), http.StatusInternalServerError)
-		return
-	}
-
-	var runtimeHookPath config.RuntimeRequestPath
-	var hookReq interface{}
-	containerMeta := store.GetContainerInfo(containerID)
-	if containerMeta != nil {
-		runtimeHookPath = config.StopContainer
-		hookReq = containerMeta.GetContainerResourceHookRequest()
-	} else {
-		// sandbox container
-		runtimeHookPath = config.StopPodSandbox
-		podInfo := store.GetPodSandboxInfo(containerID)
-		if podInfo == nil {
-			// refuse the req
-			http.Error(wr, "Failed to get pod info", http.StatusInternalServerError)
-			return
-		}
-		hookReq = podInfo.GetPodSandboxHookRequest()
+func (d *RuntimeManagerDockerServer) DockerRequestHandler(ctx context.Context, wr http.ResponseWriter, req *http.Request) {
+	resExecutor := &PodResourceExecutorDocker{
+		cgroupDriver: d.cgroupDriver,
 	}
-
-	d.Direct(wr, req)
-
-	if containerMeta != nil {
-		store.DeleteContainerInfo(containerID)
-	} else {
-		store.DeletePodSandboxInfo(containerID)
-	}
-
-	if _, err := d.dispatcher.Dispatch(ctx, runtimeHookPath, config.PostHook, hookReq); err != nil {
-		klog.Errorf("Failed to call post stop hook server %v", err)
-		http.Error(wr, err.Error(), http.StatusInternalServerError)
-		return
-	}
-}
-
-func (d *RuntimeManagerDockerServer) HandleUpdateContainer(ctx context.Context, wr http.ResponseWriter, req *http.Request) {
-	containerID, err := getContainerID(req.URL.Path)
-	if err != nil {
-		klog.Errorf("Failed to get container id, err: %v", err)
-		http.Error(wr, err.Error(), http.StatusInternalServerError)
+	if err := resExecutor.ParseRequest(req); err != nil {
+		d.generateErrorInfo(wr, err.Error())
 		return
 	}
+	defer resExecutor.DeleteCheckpointIfNeed(req)
 
-	reqBytes, err := io.ReadAll(req.Body)
-	if err != nil {
-		klog.Errorf("Failed to ready req body, err: %v", err)
-		http.Error(wr, err.Error(), http.StatusInternalServerError)
+	// no need to care about the resp
+	if resp, err := d.dispatcher.Dispatch(ctx, resExecutor.RuntimeRequestPath, config.PreHook, resExecutor.GenerateHookRequest()); err != nil {
+		d.generateErrorInfo(wr, "failed to call pre start container hook server %v", err)
 		return
-	}
-	containerConfig := &container.UpdateConfig{}
-	if err := json.Unmarshal(reqBytes, containerConfig); err != nil {
-		klog.Errorf("Failed to Unmarshal req body to docker config, err: %v", err)
-		http.Error(wr, err.Error(), http.StatusInternalServerError)
+	} else if err = resExecutor.ReConstructRequestByHookResponse(resp, req); err != nil {
+		d.generateErrorInfo(wr, "failed to reconstruct request by hook server response %v", err)
```

Comment:
I had a conversation with @ZYecho last week about the following situation where if hook gives an error response, whether proxy should directly return or simply bypass it. The reason we discuss it because we found that the docker will directly return but cri will bypass as shown by the following code: 
https://github.com/koordinator-sh/koordinator/blob/main/pkg/runtimeproxy/server/cri/criserver.go#L108 

Currently the hook failure policy is global https://github.com/koordinator-sh/koordinator/blob/main/examples/runtime-hook-server/koordlet.json#L3, so do you think it is a good idea to make the failure policy to a hook granularity? so we can check if we should bypass the specific hook when hook return an error to proxy? What do you think?